### PR TITLE
[DONT MERGE] suggestion: way of loading library from resources

### PR DESCRIPTION
**Pull request by i10416**
## Design of `LibraryName(@resourcename)`

I think `compileCachedResourceIfNeeded`(compiling native sources via clang [here](https://github.com/markehammons/slinc/compare/master...i10416:SLInC:suggestion-of-library-load?expand=1#diff-73fa61a250f08024ec6d9a6b60131be9e68de53439a359599a52b3ec0c09db58R46)) is too much and had better be delegated to build tool.

How about just copying and loading shared library for the platform where the program is running?

For example, `@LibraryName("@foo")` sends resources/native/(lib)?foo.[so|dylib|dll] to `$appDataStore/(lib)?foo.[so|dylib|dll]` and then call `System.load($appDataStore/(lib)?foo.[so|dylib|dll])` based on the platform. (e.g. libfoo.so on Linux and foo.dylib on OSX)

It seems reasonable because library authors often bundle native binary(\*.so, \*.dylib, \*.dll) for each platform in resources instead of native source files(\*.c, \*.cpp).

See 
- https://github.com/metarank/lightgbm4j/tree/main/src/main/resources 
- https://github.com/sbt/sbt-jni/blob/main/core/src/main/scala/com/github/sbt/jni/syntax/NativeLoader.scala#L33
- https://github.com/HumbleUI/JWM/blob/c40d2469c685636a9a8ee24e8300a1744859c9bd/shared/java/impl/Library.java#L40

for instance.

(note: SlinC fails to load shared library from resources now because it searches fo `$name.c`.)

I think we need to refine lookup routine here.

First, we need some lookup convention for CPU arch and platform, for example, naming convention where `@LibraryName(foo)` searches for `foo.so`, `libfoo.so`, `libfoo_x86_64.so` or `libfoo_amd64.so` on x86_64 linux machine, whereas `(lib)?foo([_arm64|aarch_64])?.[dylib|so]` on arm64 OSX machine.

[Directory-based convention used in the example above](https://github.com/metarank/lightgbm4j/tree/main/src/main/resources) would be better.

In addition, I think it would be better if we have more explicit syntax like the following. 

```
@LibraryName(darwin_amd64->foo_x86_64.dylib,darwin_arm64->foo_arm64.dylib,linux_amd64->libfoo_x86_64.so,...)
```

### CODE REVIEW

**Comment by i10416** [comment]
I want to hear your idea/feedback before refining/refactoring the routines for looking up native lib from resource dir.

File: core/src/fr/hammons/slinc/Tools.scala [hunk 1]
```diff
@@ -27,16 +27,20 @@ object Tools:
 
   def sendResourceToCache(name: String): Unit =
     Files.createDirectories(appDataStore)
-    val cacheLocation = appDataStore.resolve(s"$name.c")
-
+    val cachedSharedLibLocation = appDataStore.resolve(s"$name$sharedLibSuffix")
```
Comment (i10416) [inline core/src/fr/hammons/slinc/Tools.scala:30]: I want to pick up a shared library for the platform where program is running and send it to `appDataStore.resolve(s"$name$sharedLibSuffix")`.

File: core/src/fr/hammons/slinc/Tools.scala [hunk 2]
```diff
@@ -27,16 +27,20 @@ object Tools:
...
+    val cachedNativeSrcLocation = appDataStore.resolve(s"$name.c")
+    if !Files.exists(cachedSharedLibLocation) then
+      val stream = getClass().getResourceAsStream(s"/native/$name$sharedLibSuffix")
+      if stream != null then Files.copy(stream,cachedSharedLibLocation)
+      return
     if !Files.exists(appDataStore.resolve(s"$name.c")) then
```
Comment (i10416) [inline core/src/fr/hammons/slinc/Tools.scala:36]: I don't think it is reasonable to bundle `*.c` or `*.cpp` files in resources and compile them at runtime.

Comment (i10416) [reply core/src/fr/hammons/slinc/Tools.scala:36]: I guess current implementation is just for testing, isn't it?

**Comment by markehammons** [comment]
So, I was thinking that the "@library" syntax would be used for loading pre-build libs if they existed, or compiling C files if they didn't into a shared lib for the platform. With regards to loading the pre-built libs, I was thinking that the libs would be named as such in resources: `.<arch>.<platform>.<so|dll|dylib>`. For example, `test.amd64.linux.so` would be loaded for `@test` on linux amd64 platforms.

Re the building of C/C++ files, I'd like to have this feature for some stuff like creating shims to C++ code for slinc to call, as well as a backup in case the code is run on a new platform. Maybe it's something that I can make experimental and put behind a non-default configuration?

File: core/src/fr/hammons/slinc/Tools.scala (hunk 2 above)
Comment (markehammons) [reply core/src/fr/hammons/slinc/Tools.scala:36]: So, it's helpful for testing, but I was hoping to have this feature as a last resort for new platforms. It's not portable no, but the .c or .cpp files are supposed to be used as the last resort, and are ignored if there are .so or .dll files with the same name available. See below.

**Comment by markehammons** [comment]
Other things that need to be added to this feature is hashing the resource and checking the copied resource against it. Right now, if you update your .dll, .so, or .c, and there's already a .dll or .so copied out into the cache, your library will not be updated, the stale library will be used, and your program may well crash.

**Pull request closed without merging**